### PR TITLE
Introduce a maximum frequency for LED changes

### PR DESCRIPTION
**Pull request by PeterJCLaw**

### CODE REVIEW

File: modules/sr/robot/ruggeduino_devices.py [hunk 1]
```diff
@@ -43,6 +43,16 @@ class Led:
 
     def __init__(self, webot, device_name):
         self.webot_sensor = webot.getLED(device_name)
+        self._webot = webot
```
Comment (RealOrangeOne) [inline modules/sr/robot/ruggeduino_devices.py:46]: nit: If we're keeping reference to `webot`, might as well use it everywhere

```suggestion
        self._webot = webot
        self.webot_sensor = self._webot.getLED(device_name)
```

**Comment by prophile** [comment]
This is a maximum frequency, is it not?

**Comment by Adimote** [comment]
Tested with and without this change:

Before:
unlimited

After:
limited

File: modules/sr/robot/ruggeduino_devices.py (hunk 1 above)
Comment (PeterJCLaw) [reply modules/sr/robot/ruggeduino_devices.py:46]: I'm not sure what benefit this provides? I think either spelling is fine, so I'd rather err on the side of leaving existing code as-is.

**Comment by PeterJCLaw** [comment]
One thing which does occur to me now is that there are several LEDs on the robot, do we need to ensure that it's only four changes overall per second?

**Comment by trickeydan** [comment]
> One thing which does occur to me now is that there are several LEDs on the robot, do we need to ensure that it's only four changes overall per second?

In the interests of what this PR is for, it should limit over all LEDs inclusively.

**Comment by RealOrangeOne** [comment]
That point also raises the far more difficult implementation: Should this limit be synced between all robots?

**Comment by PeterJCLaw** [comment]
> That point also raises the far more difficult implementation: Should this limit be synced between all robots?

This would be _much_ harder to do. I suggest that we focus on getting the robot-level limit in place and come back to that separately.

File: modules/sr/robot/ruggeduino_devices.py [hunk 2]
```diff
@@ -41,18 +45,19 @@ def read_value(self):
...
-        self._last_change = 0
+        self._limiter = limiter
 
     def write_value(self, value):
-        now = self._webot.getTime()
-        diff = now - self._last_change
-        if diff < 0.6:
-            # Avoid creating strobing effects by allowing LEDs to change at most
-            # once per second.
+        if not self._limiter.can_chage():
```
Comment (trickeydan) [inline modules/sr/robot/ruggeduino_devices.py]:
```suggestion
        if not self._limiter.can_change():
```

Comment (PeterJCLaw) [reply modules/sr/robot/ruggeduino_devices.py]: I think https://github.com/srobo/competition-simulator/pull/188/commits/ba3f4cb728dc966332918a02e7a1541f4881c6b4 does this?

File: modules/sr/robot/output_frequency_limiter.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from controller import Robot
+
+# Allow up to two changes per second
+MAX_FREQUENCY = 2
```
Comment (RealOrangeOne) [inline modules/sr/robot/output_frequency_limiter.py:4]: Should the frequency be an argument to the limiter?

Comment (PeterJCLaw) [reply modules/sr/robot/output_frequency_limiter.py:4]: I thought about that but decided it would be easy enough to add that support when we needed it.

**Review by RealOrangeOne — APPROVED** [review]

**Pull request merged**